sonic_y_cable/y_cable_base.py

```diff
@@ -0,0 +1,577 @@
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+            lane:
+                 an Integer, the lane on which to collect the cursor values
+                             1 -> lane 1,
+                             2 -> lane 2
+                             3 -> lane 3
+                             4 -> lane 4
+            target:
+                 an Integer, the actual target to get the cursor values on
+                             TARGET_NIC -> NIC,
+                             TARGET_TOR1-> TOR1,
+                             TARGET_TOR2 -> TOR2
+        Returns:
+            a list, with  pre one, pre two , main, post one, post two cursor values in the order
+        """
+
+        raise NotImplementedError
+
+    def get_firmware_version(self, port, target):
+        """ This routine should return the active, inactive and next (committed)
+        firmware running on the target. Each of the version values in this context
+        could be a string with a major and minor number and a build value.
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+            target:
+                 an Integer, the actual target to get the cursor values on
+                             TARGET_NIC -> NIC,
+                             TARGET_TOR1-> TOR1,
+                             TARGET_TOR2 -> TOR2
+        Returns:
+            a Dictionary:
+                 with version_active, version_inactive and version_next keys
+                 and their corresponding values
+
+        """
+
+        raise NotImplementedError
+
+    def download_firmware(self, port, fwfile):
+        """ This routine should download and store the firmware on all the
+        components of the Y cable of the port specified.
+        This should include any internal transfers, checksum validation etc.
+        from TOR to TOR or TOR to NIC side of the firmware specified by the fwfile.
+        This basically means that the firmware which is being downloaded should be
+        available to be activated (start being utilized by the cable) once this API is
+        successfully executed.
+        Note that this API should ideally not require any rollback even if it fails
+        as this should not interfere with the existing cable functionality because
+        this has not been activated yet.
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+            fwfile:
+                 a string, a path to the file which contains the firmware image.
+                 Note that the firmware file can be in the format of the vendor's
+                 choosing (binary, archive, etc.). But note that it should be one file
+                 which contains firmware for all components of the Y-cable
+        Returns:
+            an Integer:
```

Comment:
Maybe rather than explicitly stating "An Integer", we simply state "One of the following predefined constants:". This would also allow us to change the underlying type transparently if desired.

Also, please explicitly list the acceptable constants this function should return.

Reply:
fixed